src/ser.rs

```diff
@@ -391,10 +394,25 @@ impl<'s> ser::Serializer for &'s Serializer {
 
     fn serialize_newtype_struct<T: ?Sized + Serialize>(
         self,
-        _name: &'static str,
+        name: &'static str,
         value: &T,
     ) -> Result {
-        value.serialize(self)
+        if name == PRESERVED_VALUE_MAGIC {
+            let ptr = value.serialize(self)?;
```

Comment:
Tempted to say that this would be easier & tiny bit faster by making a custom serializer that only accepts & saves `u32`, and rejects everything else... but ugh that would require a lot of boilerplate for something so small, so probably not worth it.